blueocean-core-js/src/js/components/RunButton.jsx

```diff
@@ -0,0 +1,167 @@
+/**
+ * Created by cmeyers on 8/26/16.
+ */
+
+import React, { Component, PropTypes } from 'react';
+import { Icon } from 'react-material-icons-blue';
+
+import { RunApi as runApi } from '../';
+import { SseBus as sseBus } from '../';
+import { ToastService as toastService } from '../';
+import { buildRunDetailsUrl } from '../UrlBuilder';
+
+const stopProp = (event) => {
+    event.stopPropagation();
+};
+
+/**
+ * Run Buttons allows a pipeline or branch to be run and also be stopped thereafter.
+ */
+export class RunButton extends Component {
+
+    constructor(props) {
+        super(props);
+
+        this.subscriptionId = null;
+
+        this.state = {
+            running: false,
+            stopping: false,
+        };
+    }
+
+    componentDidMount() {
+        this.subscriptionId = sseBus.subscribeToJob(
+            (runData, event) => this._onJobEvent(runData, event),
+            (event) => this._filterJob(event)
+        );
+    }
+
+    componentWillReceiveProps(nextProps) {
+        this._updateState(nextProps);
+    }
+
+    componentWillUnmount() {
+        if (this.subscriptionId) {
+            sseBus.unsubscribe(this.subscriptionId);
+        }
+    }
+
+    _updateState(nextProps) {
+        const oldStatus = this.props.latestRun && this.props.latestRun.state || '';
+        const newStatus = nextProps.latestRun && nextProps.latestRun.state || '';
+
+        // if the state of the run changed, then assume it's no longer trying to stop
+        if (oldStatus !== newStatus) {
+            this.setState({
+                stopping: false,
+            });
+        }
+    }
+
+    _onJobEvent(runData, event) {
+        const name = decodeURIComponent(
+            event.job_ismultibranch ? event.blueocean_job_branch_name : event.blueocean_job_pipeline_name
+        );
+        const runId = event.jenkins_object_id;
+
+        if (event.jenkins_event === 'job_run_started') {
+            const runDetailsUrl = buildRunDetailsUrl(runData);
+
+            toastService.newToast({
+                text: `Started "${name}" #${runId}`,
```

Comment:
A lot of duplication between these, seems like might be nice to have a more generalized component that takes a toast message, icon, and action or something.